kamon-akka-remote-2.4.x/src/main/scala/kamon/akka/instrumentation/RemotingInstrumentation.scala

```diff
@@ -1,24 +1,96 @@
 package akka.kamon.instrumentation
 
+import java.lang.reflect.Field
 import java.nio.ByteBuffer
 
-import akka.actor.{ActorRef, Address, AddressFromURIString, ExtendedActorSystem}
+import akka.actor.{ActorRef, Address, AddressFromURIString, Cell, ExtendedActorSystem}
 import akka.KamonOptionVal.OptionVal
+import akka.dispatch.sysmsg.{Failed, SystemMessage, Terminate, Watch}
 import akka.remote.WireFormats._
 import akka.remote.instrumentation.TraceContextAwareWireFormats.{AckAndTraceContextAwareEnvelopeContainer, RemoteTraceContext, TraceContextAwareRemoteEnvelope}
 import akka.remote.{Ack, RemoteActorRefProvider, SeqNo}
 import akka.util.ByteString
 import kamon.Kamon
 import kamon.akka.RemotingMetrics
+import kamon.akka.context.HasTransientContext
+import kamon.context.{HasContext, Key}
 import org.aspectj.lang.ProceedingJoinPoint
 import org.aspectj.lang.annotation._
 
+import scala.collection.mutable
 import scala.util.Try
 
+
+
+@Aspect
+class HasTransientContextIntoSend {
+  @DeclareMixin("akka.remote.EndpointManager.Send+")
+  def mixinHasContextToSystemMessage: HasContext = HasTransientContext.fromCurrentContext()
+
+  @Pointcut("execution(akka.remote.EndpointManager.Send+.new(..)) && this(send)")
+  def sendCreationCreation(send: HasContext): Unit = {}
+
+  @After("sendCreationCreation(send)")
+  def afterSystemMessageCreation(send: HasContext): Unit = {
+    // Necessary to force the initialization of HasContext at the moment of creation.
+    send.context
+  }
+
+
+  @Pointcut("execution(* akka.remote.EndpointWriter.writeSend(*)) && args(send)")
+  def writingToTransport(send: HasContext): Unit = {}
+
+  @Around("writingToTransport(send)")
+  def aroundWriteToTransport(pjp: ProceedingJoinPoint, send: HasContext) = {
+    Kamon.withContext(send.context) {
+      pjp.proceed()
+    }
+  }
+}
+
+
+object RemotingInstrumentation {
+  val contextFields: mutable.Map[Class[_], Field] = mutable.Map.empty
```

Comment:
This would need to be made thread safe by either turning into a TrieMap/etc or synchronizing access. Although, given that we know how many SystemMessages are there and the number is pretty low, why not just initialize all of them at once and use a match on the message classes instead of doing the lookup for every message?